### PR TITLE
[Spark] Fix schema evolution issue with nested struct (within a map) and column renamed

### PR DESCRIPTION
**Pull request by Richard-code-gig**
This PR fixes an issue with schema evolution in Delta Lake where adding a new field to a struct within a map and renaming an existing top level field caused the operation to fail.

The fix includes logic to handle these transformations properly, ensuring that new fields are added without conflicts.

It also resolved a ToDo of casting map types in the [DeltaAnalysis.scala](https://github.com/Richard-code-gig/delta/blob/feature/schema-evolution-with-map-fix/spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala) module.

### Changes:
- Updated schema evolution logic to support complex map transformations.
- Enabled schema evolution for both map keys, simple and nested values
- Added additional case statements to handle MapTypes in addCastToColumn method in DeltaAnalysis.scala module.
- Modified TypeWideningInsertSchemaEvolutionSuite test to support schema evolution of maps.
- Added an additional method (addCastsToMaps) to DeltaAnalysis.scala module.
- Changed argument type of addCastToColumn from attributes to namedExpression
- Added [EvolutionWithMap](https://github.com/Richard-code-gig/delta/blob/feature/schema-evolution-with-map-fix/examples/scala/src/main/scala/example/EvolutionWithMap.scala) in the example modules to demonstrate use case.
- Modified nested struct type evolution with field upcast test in map in TypeWideningInsertSchemaEvolutionSuite.scala
- Added new tests cases for maps to DeltaInsertIntoTableSuite.scala

### Related Issues:
- Resolves: #3227 

#### Which Delta project/connector is this regarding?

- [✓] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## How was this patch tested?
Tested through:
- Integration Tests: Validated changes with Delta Lake and Spark integration. See [EvolutionWithMap](https://github.com/Richard-code-gig/delta/blob/feature/schema-evolution-with-map-fix/examples/scala/src/main/scala/example/EvolutionWithMap.scala).
- Validated the test suites passed and [TypeWideningInsertSchemaEvolutionSuite](https://github.com/Richard-code-gig/delta/blob/feature/schema-evolution-with-map-fix/spark/src/test/scala/org/apache/spark/sql/delta/typewidening/TypeWideningInsertSchemaEvolutionSuite.scala) to add support for maps.
- Added additional tests cases in [DeltaInsertIntoTableSuite](https://github.com/Richard-code-gig/delta/blob/feature/schema-evolution-with-map-fix/spark/src/test/scala/org/apache/spark/sql/DeltaInsertIntoTableSuite.scala) to cover complex map transformations

## Does this PR introduce _any_ user-facing changes?
No, it doesn't introduce any user-facing changes. It only resolved an issue even in the released versions of Delta Lake.

The previous behaviour was an error message when attempting operations involving adding extra fields to StructField in maps:
[[DATATYPE_MISMATCH.CAST_WITHOUT_SUGGESTION](https://docs.databricks.com/error-messages/error-classes.html#datatype_mismatch.cast_without_suggestion)] Cannot resolve "metrics" due to data type mismatch: cannot cast "MAP<STRING, STRUCT<id: INT, value: INT, comment: STRING>>" to "MAP<STRING, STRUCT<id: INT, value: INT>>".

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala
Comment (johanl-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:1144]: "Recursively casts map data types in case the key/value type differs"

I'm not entirely sure why the struct method mentions `in case it contains null types` but I don't think that's accurate

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala
Comment (johanl-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:1065]: I would just remove the TODO at this point. Both arrays and maps are now (partially) handled. We can call out their respective limitations on the dedicated methods if needed

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala
Comment (johanl-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:1085]: I'm curious, do you have the stack trace for the error you're seeing?

It should be possible to handle arbitrary nesting of maps/arrays/structs. although that might require a pretty deep refactor of this code and a high risk of introducing regressions or breaking changes that would likely negate the benefits

File: spark/src/test/scala/org/apache/spark/sql/delta/typewidening/TypeWideningInsertSchemaEvolutionSuite.scala
Comment (johanl-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/typewidening/TypeWideningInsertSchemaEvolutionSuite.scala:300]: It would be good to add more tests, for example in [DeltaInsertIntoSQLSuite](https://github.com/delta-io/delta/blob/3a98b8a33a48416649d0844355acffceca4ae21c/spark/src/test/scala/org/apache/spark/sql/delta/DeltaInsertIntoTableSuite.scala#L44):
- add a test covering the initial issue from https://github.com/delta-io/delta/issues/3227 - essentially a version of the example you provide, except that I don't think we run examples as part of PR tests (I might be wrong)
- add tests covering inserting data into a map type using key/values that have (1) a different type than in the table, (2) more column than in the table with schema evolution enabled/disabled
- if supporting maps/arrays nested inside maps: add a test covering 2 level deep nested maps

File: spark/src/test/scala/org/apache/spark/sql/delta/typewidening/TypeWideningInsertSchemaEvolutionSuite.scala
Comment (johanl-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/typewidening/TypeWideningInsertSchemaEvolutionSuite.scala:316]: You can actually remove this test case altogether and update the one below to remove `includeInserts` so that it will cover all insert types.

**Review by johanl-db — COMMENTED** [review]
Thanks @Richard-code-gig for taking the time to work on this!

The code looks good, two main points I see to address before we can merge this:
- Support casting map keys and not just map values. It could be surprising for users otherwise to see that one works but not the other.
- Add a few more test cases, I provided some ideas.

nice to have but not required:
- support deeply nested maps
that's arguably a less common use case and I don't know if it's going to be straightforward or not so it's ok if we don't support it.

Also, ignore the Delta Spark Master job failure, it has been broken for quite some time and won't be a blocker to merge this change

**Comment by Richard-code-gig** [comment]
Thanks for the insight @johanl-db.
I will try get the stack trace for maxIteration error too

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala
Comment (Richard-code-gig) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:934]: Hi @johanl-db,
Should we be calling addCastsToColumn here especially as addCastsToColumn takes an attribute type for both parent and targetAttr arguments instead of NamedLambdaVariable.

Looking forward to your thoughts.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaInsertIntoTableSuite.scala
Comment (Richard-code-gig) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaInsertIntoTableSuite.scala]: Hi Johan,
Yea, I agree with you on the comments and suggestions. Hopefully it is ready now for merging when you're back.
Enjoy your holidays :)

**Review by johanl-db — APPROVED** [review]
Nice work @Richard-code-gig!

These aren't blocker, you can address them if you want, otherwise give a heads up to @vkorukanti who can merge the PR on your behalf

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala
Comment (Richard-code-gig) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:1155]: Welcome back Johanl-db.
I can address the style issues after work this evening.
The key.toAttribute is causing the maxIteration error I wrote about earlier.

Comment (johanl-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala:1155]: Ah right, that's probably because the data type isn't correct with `key.toAttribute` - it's the source type where it should be the target type - and that trips the analyzer

**Comment by Richard-code-gig** [comment]
> Nice work @Richard-code-gig!
> 
> These aren't blocker, you can address them if you want, otherwise give a heads up to @vkorukanti who can merge the PR on your behalf

Hi @johanl-db  @vkorukanti I've resolved the style issues. It can be merged now.

**Pull request merged**
